### DISCUSSION TITLE
Allow duplicate query parameters

### DISCUSSION THREAD

**Discussion by LuckyTurtleDev** [tokio-rs/axum#2380] [Q&A] [answered]
### Summary

How can I allow a query parameter to occur a second time? I am only interested at the first value, but the second should not result  a deserialize error.

The following request fails:
```
http://localhost:8080/?theme=dark&id=1234&theme=dark
Failed to deserialize query string: duplicate field `theme`
```
```rust
#[derive(Deserialize)]
pub struct QueryData {
    theme: Theme,
    id: u64
}
```

I have also try to use `Vec<Theme>` but this also not work:
```
Failed to deserialize query string: invalid type: string "dark", expected a sequence
```

### axum version

0.7.1

**Comment by davidpdrsn** [comment]
axum uses https://crates.io/crates/serde_urlencoded under the hood so there isn't anything we can do about that. Either open an issue with them or try https://crates.io/crates/serde_qs which also has axum support. Not sure if serde_qs supports it but worth a shot.

**Reply by anarchodin** [reply]
I don't think they'll take kindly to an issue, given [prior responses](https://github.com/nox/serde_urlencoded/issues/109). This means `serde_urlencoded` is intentionally non-conformant with the `application/x-www-form-urlencoded` spec. It seems like the `axum_extra` implementation should be the default one.

**Comment by jplatte** [comment, marked as answer]
You can use `axum_extra::extract::Query` (not the same as `axum::extract::Query`) if you change your `QueryData` definition to

```rust
#[derive(Deserialize)]
pub struct QueryData {
    #[serde(default)]
    theme: Vec<Theme>,
    id: u64
}
```

It's also possible to write a custom deserialization function that only extracts the first item from a sequence, but that's not as simple.

**Reply by LuckyTurtleDev** [reply]
Thanks `axum_extra::extract::Query` works for me:

```rust
use axum_extra::extract::Query;

#[derive(Deserialize)]
pub struct QueryData {
    #[serde(default, deserialize_with = "take_first")]
    theme: Theme,
    id: u64
}

pub fn take_first<'de, D, T>(deserializer: D) -> Result<T, D::Error>
where
	D: de::Deserializer<'de>,
	T: Deserialize<'de> + Default
{
	let vec: Vec<T> = Vec::deserialize(deserializer)?;
	Ok(vec.into_iter().next().unwrap_or_default())
}
```


or with `Option`:
```rust
use axum_extra::extract::Query;

#[derive(Deserialize)]
pub struct QueryData {
    #[serde(default, deserialize_with = "take_first")]
    theme: Option<Theme>,
    id: u64
}

pub fn take_first<'de, D, T>(deserializer: D) -> Result<Option<T>, D::Error>
where
	D: de::Deserializer<'de>,
	T: Deserialize<'de>
{
	let vec: Vec<T> = Vec::deserialize(deserializer)?;
	Ok(vec.into_iter().next())
}
```

**Reply by jplatte** [reply]
Note that `serde(default)` means that if the parameter is not specified, the `Default` impl for `Theme` will be used. I usually include this in examples of `Vec` usage with `serde_html_form` / `axum_extra::extract::Query` because you often want to allow zero occurrences to be supported as well. You can leave out `serde(default)` if the parameter must be specified at least once.

**Reply by LuckyTurtleDev** [reply]
> Because you often want to allow zero occurrences to be supported as well.

If I do not use `default` in the first example, the vector is not empty instead I get an error that `Theme` is missing.
```
Failed to deserialize query string: missing field `theme`
```

**Reply by jplatte** [reply]
Yes, exactly. If you *want* to enforce that it's specified at least once, don't use `serde(default)`.
